File: benchmarks/livermore/k18_hydro_2d.c
/* K18 — 2-D Explicit Hydrodynamics (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION ZA(101,7), ZP(101,7), ZQ(101,7), ZR(101,7),
            ZM(101,7), ZB(101,7), ZU(101,7), ZV(101,7), ZZ(101,7)
   KN=6, JN=100. Column-major: arr(J,K) at flat (K-1)*101 + J (1-based) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define JN  100
#define KN  6
#define JD  101
#define KD  7
#define NREPS 1382000

/* 1-based column-major: arr(j,k) with first dim 101 */
#define ZA(j,k) za[((k)-1)*101 + (j)]
#define ZP(j,k) zp[((k)-1)*101 + (j)]
#define ZQ(j,k) zq[((k)-1)*101 + (j)]
#define ZR(j,k) zr[((k)-1)*101 + (j)]
#define ZM(j,k) zm[((k)-1)*101 + (j)]
#define ZB(j,k) zb[((k)-1)*101 + (j)]
#define ZU(j,k) zu[((k)-1)*101 + (j)]
#define ZV(j,k) zv[((k)-1)*101 + (j)]
#define ZZ(j,k) zz[((k)-1)*101 + (j)]

int main(void) {
    /* 101*7=707, max flat index 707, need [708] */
    double za[708], zp[708], zq[708], zr[708];
    double zm[708], zb[708], zu[708], zv[708], zz[708];

    /* Initialise arrays (1-based) */
    signel(&zp[1], JD * KD);
    signel(&zq[1], JD * KD);
    signel(&zr[1], JD * KD);
    signel(&zm[1], JD * KD);
    /* Ensure zm large enough for division */
    for (long i = 1; i <= JD * KD; i++) zm[i] += 10.0;
    signel(&zz[1], JD * KD);
    /* Zero out za, zb, zu, zv */
    for (long k = 1; k <= KD; k++)
        for (long j = 1; j <= JD; j++) {
            ZA(j,k) = 0.0; ZB(j,k) = 0.0;
            ZU(j,k) = 0.0; ZV(j,k) = 0.0;
        }

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double t, s;
    for (long l = 0; l < NREPS; l++) {
        t = 0.0037; s = 0.0041;

        /* Phase 1: compute ZA, ZB */
        for (long k = 2; k <= KN; k++)
            for (long j = 2; j <= JN; j++) {
                ZA(j,k) = (ZP(j-1,k+1)+ZQ(j-1,k+1)-ZP(j-1,k)-ZQ(j-1,k))
                          *(ZR(j,k)+ZR(j-1,k))
                          /(ZM(j-1,k)+ZM(j-1,k+1));
                ZB(j,k) = (ZP(j-1,k)+ZQ(j-1,k)-ZP(j,k)-ZQ(j,k))
                          *(ZR(j,k)+ZR(j,k-1))
                          /(ZM(j,k)+ZM(j-1,k));
            }

        /* Phase 2: update ZU, ZV */
        for (long k = 2; k <= KN; k++)
            for (long j = 2; j <= JN; j++) {
                ZU(j,k) = ZU(j,k)
                  + s*(ZA(j,k)*(ZZ(j,k)-ZZ(j+1,k))
                      -ZA(j-1,k)*(ZZ(j,k)-ZZ(j-1,k))
                      -ZB(j,k)*(ZZ(j,k)-ZZ(j,k-1))
                      +ZB(j,k+1)*(ZZ(j,k)-ZZ(j,k+1)));
                ZV(j,k) = ZV(j,k)
                  + s*(ZA(j,k)*(ZR(j,k)-ZR(j+1,k))
                      -ZA(j-1,k)*(ZR(j,k)-ZR(j-1,k))
                      -ZB(j,k)*(ZR(j,k)-ZR(j,k-1))
                      +ZB(j,k+1)*(ZR(j,k)-ZR(j,k+1)));
            }

        /* Phase 3: update ZR, ZZ */
        for (long k = 2; k <= KN; k++)
            for (long j = 2; j <= JN; j++) {
                ZR(j,k) = ZR(j,k) + t*ZU(j,k);
                ZZ(j,k) = ZZ(j,k) + t*ZV(j,k);
            }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("zu(51,4) = %f\n", ZU(51,4));
    return 0;
}

File: benchmarks/livermore/k18_hydro_2d.f
C     Kernel 18 -- 2-D Explicit Hydrodynamics
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K18
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION ZA(101,7), ZP(101,7), ZQ(101,7), ZR(101,7)
      DIMENSION ZM(101,7), ZB(101,7), ZU(101,7), ZV(101,7), ZZ(101,7)
C
C     Initialize ZP using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      SCALED = 0.1D0
      DO 12 J = 1, 7
         DO 11 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZP(I,J) = (BUZZ - FIZZ) * SCALED
   11    CONTINUE
   12 CONTINUE
C
C     Initialize ZQ using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 14 J = 1, 7
         DO 13 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZQ(I,J) = (BUZZ - FIZZ) * SCALED
   13    CONTINUE
   14 CONTINUE
C
C     Initialize ZR using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 16 J = 1, 7
         DO 15 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZR(I,J) = (BUZZ - FIZZ) * SCALED
   15    CONTINUE
   16 CONTINUE
C
C     Initialize ZM using SIGNEL (add 10.0 to ensure large enough for division)
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 18 J = 1, 7
         DO 17 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZM(I,J) = (BUZZ - FIZZ) * SCALED + 10.0D0
   17    CONTINUE
   18 CONTINUE
C
C     Initialize ZZ using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 22 J = 1, 7
         DO 21 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZZ(I,J) = (BUZZ - FIZZ) * SCALED
   21    CONTINUE
   22 CONTINUE
C
C     Zero out ZA, ZB, ZU, ZV
C
      DO 26 J = 1, 7
         DO 25 I = 1, 101
            ZA(I,J) = 0.0D0
            ZB(I,J) = 0.0D0
            ZU(I,J) = 0.0D0
            ZV(I,J) = 0.0D0
   25    CONTINUE
   26 CONTINUE
C
C     Kernel 18: 2-D Explicit Hydrodynamics
C     C code uses 0-based [k][j] with k=row, j=col.
C     In Fortran 1-based: ZA(J,K) etc. with J=col=2..JN, K=row=2..KN
C
      DO 500 IREP = 1, 10000
         T  = 0.0037D0
         S  = 0.0041D0
         KN = 6
         JN = 100
C
C        First loop: compute ZA and ZB
         DO 120 K = 2, KN
            DO 110 J = 2, JN
               ZA(J,K) = (ZP(J-1,K+1)+ZQ(J-1,K+1)
     &                   -ZP(J-1,K)-ZQ(J-1,K))
     &                  *(ZR(J,K)+ZR(J-1,K))
     &                  /(ZM(J-1,K)+ZM(J-1,K+1))
               ZB(J,K) = (ZP(J-1,K)+ZQ(J-1,K)
     &                   -ZP(J,K)-ZQ(J,K))
     &                  *(ZR(J,K)+ZR(J,K-1))
     &                  /(ZM(J,K)+ZM(J-1,K))
  110       CONTINUE
  120    CONTINUE
C
C        Second loop: update ZU and ZV
         DO 220 K = 2, KN
            DO 210 J = 2, JN
               ZU(J,K) = ZU(J,K)
     &           + S*(ZA(J,K)*(ZZ(J,K)-ZZ(J+1,K))
     &               -ZA(J-1,K)*(ZZ(J,K)-ZZ(J-1,K))
     &               -ZB(J,K)*(ZZ(J,K)-ZZ(J,K-1))
     &               +ZB(J,K+1)*(ZZ(J,K)-ZZ(J,K+1)))
               ZV(J,K) = ZV(J,K)
     &           + S*(ZA(J,K)*(ZR(J,K)-ZR(J+1,K))
     &               -ZA(J-1,K)*(ZR(J,K)-ZR(J-1,K))
     &               -ZB(J,K)*(ZR(J,K)-ZR(J,K-1))
     &               +ZB(J,K+1)*(ZR(J,K)-ZR(J,K+1)))
  210       CONTINUE
  220    CONTINUE
C
C        Third loop: update ZR and ZZ
         DO 320 K = 2, KN
            DO 310 J = 2, JN
               ZR(J,K) = ZR(J,K) + T*ZU(J,K)
               ZZ(J,K) = ZZ(J,K) + T*ZV(J,K)
  310       CONTINUE
  320    CONTINUE
  500 CONTINUE
C
      WRITE(*,900) ZU(51,4)
  900 FORMAT('K18 hydro 2D: zu(51,4) = ', E23.15E3)
      END
